src/explore-education-statistics-admin/src/pages/methodology/edit-methodology/content/context/MethodologyContext.tsx

```diff
@@ -0,0 +1,176 @@
+import remove from 'lodash/remove';
+import React, { createContext, ReactNode, useContext } from 'react';
+import { MethodologyContent } from 'src/services/methodology/types';
+import { Reducer, useImmerReducer } from 'use-immer';
+import MethodologyDispatchAction from './MethodologyContextActionTypes';
+
+type MethodologyContextDispatch = (action: MethodologyDispatchAction) => void;
+type MethodologyContextState = {
+  methodology: MethodologyContent | undefined;
+  canUpdateMethodology: boolean;
+};
+type MethodologyProviderProps = { children: ReactNode };
+
+const MethodologyStateContext = createContext<
+  MethodologyContextState | undefined
+>(undefined);
+const MethodologyDispatchContext = createContext<
+  MethodologyContextDispatch | undefined
+>(undefined);
+
+export const methodologyReducer: Reducer<
+  MethodologyContextState,
+  MethodologyDispatchAction
+> = (draft, action) => {
+  switch (action.type) {
+    case 'CLEAR_STATE': {
+      return {
+        methodology: undefined,
+        canUpdateMethodology: false,
+      };
+    }
+    case 'SET_STATE': {
+      return { ...draft, ...action.payload };
+    }
+    case 'REMOVE_BLOCK_FROM_SECTION': {
+      const { sectionId, blockId, sectionKey } = action.payload.meta;
+      if (!draft.methodology?.[sectionKey]) {
+        throw new Error('REMOVE_BLOCK_FROM_SECTION: failed');
+      }
+
+      const matchingSection = draft.methodology[sectionKey].find(
+        section => section.id === sectionId,
+      );
+      if (matchingSection?.content) {
+        remove(matchingSection.content, content => content.id === blockId);
+      }
+
+      return draft;
+    }
+    case 'UPDATE_BLOCK_FROM_SECTION': {
+      const { block, meta } = action.payload;
+      const { sectionId, blockId, sectionKey } = meta;
+      if (!draft.methodology?.[sectionKey]) {
+        throw new Error('UPDATE_BLOCK_FROM_SECTION: failed');
+      }
+      const matchingSection = draft.methodology[sectionKey].find(
+        section => section.id === sectionId,
+      );
+      if (matchingSection?.content) {
+        const blockIndex = matchingSection.content.findIndex(
+          contentBlock => contentBlock.id === blockId,
+        );
+        matchingSection.content[blockIndex] = block;
+      }
+
+      return draft;
+    }
+    case 'ADD_BLOCK_TO_SECTION': {
+      const { block, meta } = action.payload;
+      const { sectionId, sectionKey } = meta;
+      if (!draft.methodology?.[sectionKey]) {
+        throw new Error('ADD_BLOCK_TO_SECTION: failed');
+      }
+
+      // .comments needs initialising to array as will be undefined if empty
+      const newBlock = { ...block, comments: block.comments || [] };
```

Comment:
Why even let the `block.comments` be undefined? Could just ensure that there is a default empty array at the source

Reply:
methodology doesn't allow comments anyway, so I've removed this